examples/raja-teams.cpp

```diff
@@ -0,0 +1,361 @@
+#else
+using teams0 = RAJA::LoopPolicy<RAJA::loop_exec, RAJA::loop_exec>;
+using teams1 = RAJA::LoopPolicy<RAJA::loop_exec, RAJA::loop_exec>;
+#endif
+
+
+#ifdef RAJA_ENABLE_CUDA
+using threads0 = RAJA::LoopPolicy<RAJA::loop_exec, RAJA::cuda_thread_x_loop>;
+using threads1 = RAJA::LoopPolicy<RAJA::loop_exec, RAJA::cuda_thread_y_loop>;
+#else
+using threads0 = RAJA::LoopPolicy<RAJA::loop_exec, RAJA::loop_exec>;
+using threads1 = RAJA::LoopPolicy<RAJA::loop_exec, RAJA::loop_exec>;
+#endif
+
+
+int main()
+{
+
+  // N is number of blocks in each matrix
+  const int NBlocks = 4;
+#ifdef RAJA_ENABLE_CUDA
+  const int NThreads = CUDA_BLOCK_SIZE;
+  const int N = NThreads * NBlocks;
+#else
+  const int NThreads = 1;
+  const int N = NThreads * NBlocks;
+#endif
+
+  //
+  // Allocate and initialize matrix data.
+  //
+  double *A = memoryManager::allocate<double>(N * N);
+  double *B = memoryManager::allocate<double>(N * N);
+  double *C = memoryManager::allocate<double>(N * N);
+
+  for (int row = 0; row < N; ++row) {
+    for (int col = 0; col < N; ++col) {
+      A(row, col) = row;
+      B(row, col) = col;
+    }
+  }
+
+  std::cout << "\n Running RAJA-Teams V2-version of matrix multiplication...\n";
+  std::cout << "  N = " << N << std::endl;
+
+  for (int exec_place = 0; exec_place < (int)RAJA::NUM_PLACES; ++exec_place) {
+    RAJA::ExecPlace select_cpu_or_gpu = (RAJA::ExecPlace)exec_place;
+    // auto select_cpu_or_gpu = RAJA::HOST;
+    // auto select_cpu_or_gpu = RAJA::DEVICE;
+
+    /*
+     * launch just starts a "kernel" it's doesn't provide any looping.
+     *
+     * The first argument determines which policy should be executed,
+     *
+     * The second argument is the number of teams+threads needed for each of the
+     * policies.
+     *
+     * Third argument is the lambda for the policy.
+     *
+     *
+     * The lambda takes a "resource" object, which has the teams+threads and
+     * policy selection information.
+     */
+
+    //========================
+    // Upper triangular pattern
+    //========================
+    const int N_tri = 5;
+    RAJA::launch<launch_policy>(
+        select_cpu_or_gpu,
+        RAJA::ResourceList{
+            RAJA::Resources(RAJA::Threads(N_tri)),
+            RAJA::Resources(RAJA::Teams(N_tri), RAJA::Threads(N_tri))},
+        [=] RAJA_HOST_DEVICE(RAJA::LaunchContext ctx) {
+          RAJA::loop<teams0>(ctx, RAJA::RangeSegment(0, N_tri), [=](int i) {
+            // do a matrix triangular pattern
+            RAJA::loop<threads0>(ctx, RAJA::RangeSegment(i, N_tri), [=](int j) {
+              printf("i=%d, j=%d\n", i, j);
+            });  // loop j
+          });    // loop i
+        });      // kernel
+
+    //========================
+    // Matrix-Matrix Multiplication Example
+    //========================
+    // Set up Teams/Threads
+
+    RAJA::launch<launch_policy>(select_cpu_or_gpu,
+          RAJA::ResourceList{
+            RAJA::Resources(RAJA::Threads(NBlocks, NBlocks)),
```

Comment:
@ajkunen In a sequential setting what do RAJA::Threads(...) do?

Reply:
Nothing! absolutely Nothing!

I guess that should really be an error?!?   I figured on the host, that threads would translate into OpenMP threads if someone used an OpenMP policy.

Reply:
also, I've never really looked at omp teams on the host... how does that map?

Reply:
I think we can just have users always specify teams and threads, but have it do nothing for sequential and OpenMP policies. 

For OpenMP, a possible solution is to treat the outer loops (or team loops) as acting in parallel of each other, and inner loops as acting serially (in a perfect world, inner loops could carryout SIMD instructions). 

The standard C code would look like this: 

```
#pragma omp parallel for 
for (int i=0; i<TeamBounds; ++i){

TEAM_SHARED s_A[] 

for (int j=0; j<ThreadBounds; ++j){
 /// Code that could vectorize
}

}
```

for our code code it would translate to:

```
 RAJA::loop<teams>(ctx, RAJA::RangeSegment(0, TeamBounds), [&](int bx) { 

TEAM_SHARED s_A[]; 

 RAJA::loop<threads>(ctx, RAJA::RangeSegment(0, NThreads), [&](int tx) {
 /// Code that could vectorize
});
});
```

the omp policy for teams would be implemented as:

```
template <typename SEGMENT>
struct LoopExecute<omp_loop, SEGMENT> {

  template <typename BODY>
  static RAJA_INLINE RAJA_DEVICE void exec(LaunchContext const &ctx,
                               SEGMENT const &segment,
                               BODY const &body)
  {

    int len = segment.end() - segment.begin();
   #pragma omp parallel for
    for (int i = 0; i < len; i++) {
      body(*(segment.begin() + i));
    }
  }
};
```

If we have multiple levels of outer loops we can add a RAJA::loop method that takes multiple range segments and the implementation would add a collapse clause 

```
#pragma omp parallel for collapse(3) 
for (int bz= 0; bz< zlen; bz++) 
for (int by= 0; by< ylen; by++) 
for (int bx= 0; bx< xlen; bx++) 
```

Reply:
largely agree with @artv3 here. OpenMP actually supports teams on the host now, but it's early enough in terms of implementation support that it's unlikely to buy us anything.  At some future point though you could imagine it doing something like putting teams on different numa nodes, and giving them separate contention groups so their barriers and locks and so-forth have less cache impact.  🤷 

Either way, I'd probably argue for having all three levels work regardless of the target they're given.  This could be a really good excuse to finally do the more abstract "run somehow on OpenMP" and "run somehow on CUDA" and so-forth policies we've had people request a few times.  The nesting gives some control, but also gives us some if the target selection is relatively abstract.

Reply:
I'll be taking a swing at the OpenMP (CPU) backend this week. Once I have some preliminary results I'll reach out and share what I see.